### PR TITLE
implement GenericDataFrame for DeduplicateAction and HistorizeAction

### PR DESCRIPTION
**Pull request by zzeekk**
see #916

### CODE REVIEW

File: sdl-core/src/main/scala/io/smartdatalake/definitions/Environment.scala [hunk 1]
```diff
@@ -519,6 +521,50 @@ object Environment extends SmartDataLakeLogger {
   }
   var _analyzeTableColumnMaxBytesThreshold: Option[Int] = None
 
+  /**
+   * Threshold for analyzing table columns.
+   * If table size is bigger than analyzeTableColumnMaxThresholdBytes, table columns are not analyzed.
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/definitions/Environment.scala]: It looks like this comment has nothing to do with this method

File: sdl-core/src/main/scala/io/smartdatalake/util/evolution/SchemaEvolution.scala [hunk 2]
```diff
@@ -230,17 +201,36 @@ object SchemaEvolution extends SmartDataLakeLogger {
...
-      logger.info(s"new schema: ${newDf.schema.treeString}")
+      val infoList = tgtColumns.flatMap(_.infoMsg).map("-> " + _).mkString("\n")
+      val infoTxt = s"$infoList\nold schema:\n${oldDf.schema.treeString().stripTrailing()}\nnew schema:\n${newDf.schema.treeString().stripTrailing()}".indent(2)
+      logger.info(s"schema evolution needed. mapping is:\n$infoTxt"
+      )
 
       // prepare dataframes
-      val oldExtendedDf = oldDf.select(tgtColumns.flatMap(_.oldToNewColumn):_*)
-      val newExtendedDf = newDf.select(tgtColumns.flatMap(_.newColumn):_*)
+      val oldExtendedDf = oldDf.select(tgtColumns.flatMap(_.oldToNewColumn))
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/util/evolution/SchemaEvolution.scala:210]: what's the reason for removing :_* ?

File: sdl-core/src/main/scala/io/smartdatalake/util/misc/SchemaUtil.scala [hunk 3]
```diff
@@ -79,6 +93,8 @@ object SchemaUtil {
    *
    * @param ignoreNullable whether to ignore differences in nullability.
    * @return The set of fields in `right` that are not contained in `left`.
+   *
+   *         TODO: probably doesnt work for structs nested in arrays...
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/util/misc/SchemaUtil.scala]: shall we create a ticket for that?

File: sdl-core/src/main/scala/io/smartdatalake/util/spark/evolution/ValueProjector.scala [hunk 4]
```diff
@@ -88,6 +92,7 @@ private[smartdatalake] object ValueProjector {
       case (_: ShortType, _: IntegerType) => (x => x.asInstanceOf[Short].toInt)
       case (_: ByteType, _: ShortType) => (x => x.asInstanceOf[Byte].toShort)
       case (_: FloatType, _: DoubleType) => (x => x.asInstanceOf[Float].toDouble)
+      // TODO: make non-strict, e.g converting decimal 255 to byte. Also for lines below...
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/util/spark/evolution/ValueProjector.scala]: also make a ticket?

File: sdl-core/src/main/scala/io/smartdatalake/workflow/action/HistorizeAction.scala [hunk 5]
```diff
@@ -140,24 +146,24 @@ case class HistorizeAction(
...
-      val additionalMergePredicate = Some((s"existing.${Historization.historizeDummyColName} = new.${Historization.historizeDummyColName} AND timestamp'$sqlReferenceTimestamp' between existing.${TechnicalTableColumn.captured} AND existing.${TechnicalTableColumn.delimited}" +: mergeModeAdditionalJoinPredicate.toSeq).reduce(_ + " and " + _))
-      //val additionalMergePredicate = Some((s"timestamp'$sqlReferenceTimestamp' between existing.${TechnicalTableColumn.captured} AND existing.${TechnicalTableColumn.delimited}" +: mergeModeAdditionalJoinPredicate.toSeq).reduce(_ + " AND " + _))
+      val additionalMergePredicate = Some((s"existing.${Historization.historizeDummyColName} = new.${Historization.historizeDummyColName} AND timestamp'$sqlReferenceTimestamp' between existing.${Environment.capturedColumnName} AND existing.${Environment.delimitedColumnName}" +: mergeModeAdditionalJoinPredicate.toSeq).reduce(_ + " and " + _))
+      //val additionalMergePredicate = Some((s"timestamp'$sqlReferenceTimestamp' between existing.${Environment.capturedColumnName} AND existing.${Environment.delimitedColumnName}" +: mergeModeAdditionalJoinPredicate.toSeq).reduce(_ + " AND " + _))
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/action/HistorizeAction.scala]: what to do with this line?

File: sdl-core/src/test/scala/io/smartdatalake/workflow/ActionDAGRunTest.scala [hunk 6]
```diff
@@ -42,6 +42,8 @@ class ActionDAGRunTest extends FunSuite {
   private val sdlbVersionInfo = BuildVersionInfo.sdlbVersionInfo.map(_.entries().toMap)
   private val appVersionInfo = BuildVersionInfo.appVersionInfo.map(_.entries().toMap).orElse(AppUtil.getManifestVersion.map(v => Map("version"->v)))
 
+  session.catalog.listCatalogs().show
```
Comment (Geheiner) [inline sdl-core/src/test/scala/io/smartdatalake/workflow/ActionDAGRunTest.scala]: was this left here on purpose?

File: sdl-snowflake/src/main/scala/io/smartdatalake/workflow/dataframe/snowflake/SnowparkSubFeed.scala [hunk 7]
```diff
@@ -231,6 +253,14 @@ object SnowparkSubFeed extends DataFrameSubFeedCompanion with SmartDataLakeLogge
     }
   }
 
+  override def hash(column: GenericColumn): GenericColumn = {
+    val udfRaiseError = functions.udf((msg: String) => throw new RuntimeException(msg)) // Spark raise_error functions also creates a RuntimeException
```
Comment (Geheiner) [inline sdl-snowflake/src/main/scala/io/smartdatalake/workflow/dataframe/snowflake/SnowparkSubFeed.scala]: are we missing some logic here? the error is never used

**Review by Geheiner — APPROVED** [review]
Some of the TODOs and open points should be discussed, but overall it looks really impressive!

File: sdl-core/src/main/scala/io/smartdatalake/util/evolution/SchemaEvolution.scala (hunk 2 above)
Comment (zzeekk) [reply sdl-core/src/main/scala/io/smartdatalake/util/evolution/SchemaEvolution.scala:210]: The select method from GenericDataFrame takes directly a Seq().

File: sdl-core/src/main/scala/io/smartdatalake/definitions/Environment.scala (hunk 1 above)
Comment (zzeekk) [reply sdl-core/src/main/scala/io/smartdatalake/definitions/Environment.scala]: True, i'll fix it.

File: sdl-core/src/main/scala/io/smartdatalake/workflow/action/HistorizeAction.scala (hunk 5 above)
Comment (zzeekk) [reply sdl-core/src/main/scala/io/smartdatalake/workflow/action/HistorizeAction.scala]: I'll remove it.

File: sdl-core/src/test/scala/io/smartdatalake/workflow/ActionDAGRunTest.scala (hunk 6 above)
Comment (zzeekk) [reply sdl-core/src/test/scala/io/smartdatalake/workflow/ActionDAGRunTest.scala]: nope, removed.

File: sdl-snowflake/src/main/scala/io/smartdatalake/workflow/dataframe/snowflake/SnowparkSubFeed.scala (hunk 7 above)
Comment (zzeekk) [reply sdl-snowflake/src/main/scala/io/smartdatalake/workflow/dataframe/snowflake/SnowparkSubFeed.scala]: Yes, line can be removed.

File: sdl-core/src/main/scala/io/smartdatalake/util/misc/SchemaUtil.scala (hunk 3 above)
Comment (zzeekk) [reply sdl-core/src/main/scala/io/smartdatalake/util/misc/SchemaUtil.scala]: Created #935

File: sdl-core/src/main/scala/io/smartdatalake/util/spark/evolution/ValueProjector.scala (hunk 4 above)
Comment (zzeekk) [reply sdl-core/src/main/scala/io/smartdatalake/util/spark/evolution/ValueProjector.scala]: Yes, created #936

**Comment by zzeekk** [comment]
Thanks a lot @Geheiner !

**Pull request merged**